lib/publiccloud/instance.pm

```diff
@@ -333,73 +333,143 @@ sub wait_for_guestregister {
-    # Exclude 'mr_test/saptune' test case as it will introduce random softreboot failures.
-    if (!get_var('PUBLIC_CLOUD_SLES4SAP')) {
-        $self->run_ssh_command(cmd => 'sudo journalctl -b', proceed_on_failure => 1, username => $args{username}, timeout => 360);
-    }
+    # Looping until SSH port 22 is reachable or timeout.
+    while (($duration = time() - $start_time) < $args{timeout}) {
+        $exit_code = script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1);
+        last if (isok($exit_code) and not $args{wait_stop});    # ssh port open ok
+        last if (not isok($exit_code) and $args{wait_stop});    # ssh port closed ok
+        sleep 1;
+    }    # endloop
 
-    unless ($args{proceed_on_failure}) {
-        my $error_msg;
-        if ($check_port) {
-            $error_msg = sprintf("Unable to reach SSH port of instance %s with public IP:%s within %d seconds", $self->{instance_id}, $self->{public_ip},
-                $args{timeout});
-        }
-        else {
-            $error_msg = sprintf("Can not reach systemd target on instance %s with public IP:%s within %d seconds",
-                $self->{instance_id}, $self->{public_ip}, $args{timeout});
-        }
-        croak($error_msg);
+    # exit_code is 0 when shell script is ok
+    if (isok($exit_code)) {
+        $sshout = "SSH port is open\n";
     }
+    else {
+        $sshout = "SSH port is not open failed access\n";
+        $sshout .= "as expected by stopping: OK.\n" if $args{wait_stop};
+    }    # endif
+
+    # check also remote system is running, looping by --wait option:
+    if ($args{systemup_check} and isok($exit_code)) {
+        script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts")
+          if (get_var('PUBLIC_CLOUD_PERF_COLLECT') or get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME') or get_var('PUBLIC_CLOUD_CHECK_REBOOT'));
+        while (($duration = time() - $start_time) < $args{timeout}) {
+            # On boottime test we do hard reboot which may change the instance address:
+            # timeout recalculated removing consumed time until now
+            $sysout = $self->ssh_script_output(cmd => 'sudo systemctl is-system-running',
+                timeout => $args{timeout} - $duration, proceed_on_failure => 1, username => $args{username});
+            # result check
+            if ($sysout =~ m/Permission denied \(publickey\).*/) {
+                $exit_code = 2;
+                last unless $args{ignore_wrong_pubkey};
+            }
+            elsif ($sysout =~ m/initializing|starting/) {    # still starting
+                $exit_code = undef;
+            }
+            elsif ($sysout =~ m/running/) {    # startup OK
+                $exit_code = 0;
+                $sysout .= "\nSystem successfully booted";
+                last;
+            }
+            elsif ($sysout =~ m/degraded/) {    # up but with failed services to collect
```

Comment:
I think all of those `elsif` are not needed. We just want to check if the system is up. Additional checks (e.g. failing services) are not meant to take place at this point.

Reply:
no, please, those checks do are needed, because the phases are different if before , during or after. As it is now, it works well, I already did many different tests